opentracing/trace_context.go

```diff
@@ -0,0 +1,164 @@
+
+	// Converts the encoded ASCII data (see `TraceContextID.SerializeASCII()`)
+	// into a TraceContextID of the same type as returned by
+	// NewRootTraceContextID().
+	DeserializeASCIITraceContextID(encoded string) (TraceContextID, error)
+}
+
+// A `TraceContext` builds off of an implementation-provided TraceContextID and
+// adds a simple string map of "trace tags". The trace tags are special in that
+// they are propagated *in-band*, presumably alongside application data and the
+// `TraceContextID` proper. See the documentation for `SetTraceTag()` for more
+// details and some important caveats.
+//
+// Note that the `TraceContext` is managed internally by the opentracer system;
+// opentracer implementations only need to concern themselves with the
+// `TraceContextID` (which does not know about trace tags).
+type TraceContext struct {
+	Id TraceContextID
+
+	tagLock   sync.RWMutex
+	traceTags map[string]string
+}
+
+// `tags` may be nil.
+func newTraceContext(id TraceContextID, tags map[string]string) *TraceContext {
+	if tags == nil {
+		tags = map[string]string{}
+	}
+	return &TraceContext{
+		Id:        id,
+		traceTags: tags,
+	}
+}
+
+// Set a tag on this TraceContext that also propagates to future TraceContext
+// children per `NewChild()`.
+//
+// `SetTraceTag()` enables powerful functionality given a full-stack
+// opentracing integration (e.g., arbitrary application data from a mobile app
+// can make it, transparently, all the way into the depths of a storage
+// system), and with it some powerful costs: use this feature with care.
+//
+// IMPORTANT NOTE #1: `SetTraceTag()` will only propagate trace tags to
+// *future* children of the `TraceContext` (see `NewChild()`) and/or the `Span`
+// that references it.
+//
+// IMPORTANT NOTE #2: Use this thoughtfully and with care. Every key and value
+// is copied into every local *and remote* child of this `TraceContext`, and
+// that can add up to a lot of network and cpu overhead for large strings.
+//
+// Returns a reference to this TraceContext for chaining, etc.
+func (t *TraceContext) SetTraceTag(key, value string) *TraceContext {
+	t.tagLock.Lock()
+	defer t.tagLock.Unlock()
+
+	t.traceTags[key] = value
+	return t
+}
+
+// Gets the value for a trace tag given its key. Returns the empty string if
+// the value isn't found in this TraceContext.
+func (t *TraceContext) TraceTag(key string) string {
+	t.tagLock.RLock()
+	defer t.tagLock.RUnlock()
+
+	return t.traceTags[key]
+}
+
+func (t *TraceContext) NewChild() (childCtx *TraceContext, childSpanTags Tags) {
+	t.tagLock.RLock()
+	newTags := make(map[string]string, len(t.traceTags))
+	for k, v := range t.traceTags {
+		newTags[k] = v
+	}
+	t.tagLock.RUnlock()
+
+	ctxId, childSpanTags := t.Id.NewChild()
+	return &TraceContext{
+		Id:        ctxId,
+		traceTags: newTags,
+	}, childSpanTags
+}
+
+func (t *TraceContext) SerializeASCII() string {
+	// XXX: implement correctly if we like this API
```

Comment:
@bensigelman 
A few people responded to my document with strong preference against opaque strings (one or chopped up), and in favor of one header per meta tag, e.g. 
```
UberCtx-MyTag1: myvalue1
UberCtx-MyTag2: myvalue2
```
Which is, incidentally, what I've implemented in the Python API. My only concern with this approach is the encoding of tag names and values, they need to be "safe" as http headers, but for now I left it as the responsibility of the instrumentation code.  Also, the tag names have to be case-insensitive in this approach.